src/SonarQube.Client/Api/SonarQubeService.cs

```diff
@@ -0,0 +1,252 @@
+            if (userAgent == null)
+            {
+                throw new ArgumentNullException(nameof(userAgent));
+            }
+            this.messageHandler = messageHandler;
+            this.requestFactory = requestFactory;
+            this.userAgent = userAgent;
+        }
+
+        /// <summary>
+        /// Convenience overload for requests that do not need configuration.
+        /// </summary>
+        private Task<TResponse> InvokeRequestAsync<TRequest, TResponse>(CancellationToken token)
+            where TRequest : IRequest<TResponse>
+        {
+            return InvokeRequestAsync<TRequest, TResponse>(request => { }, token);
+        }
+
+        /// <summary>
+        /// Creates a new instance of the specified TRequest request, configures and invokes it and returns its response.
+        /// </summary>
+        /// <typeparam name="TRequest">The request interface to invoke.</typeparam>
+        /// <typeparam name="TResponse">The type of the request response result.</typeparam>
+        /// <param name="configure">Action that configures a type instance that implements TRequest.</param>
+        /// <param name="token">Cancellation token.</param>
+        /// <returns>Returns the result of the request invocation.</returns>
+        private async Task<TResponse> InvokeRequestAsync<TRequest, TResponse>(Action<TRequest> configure,
+            CancellationToken token)
+            where TRequest : IRequest<TResponse>
+        {
+            EnsureIsConnected();
+
+            var request = requestFactory.Create<TRequest>(sonarQubeVersion);
+            configure(request);
+
+            var result = await request.InvokeAsync(httpClient, token);
+
+            return result;
+        }
+
+        public async Task ConnectAsync(ConnectionInformation connection, CancellationToken token)
+        {
+            httpClient = new HttpClient(messageHandler)
+            {
+                BaseAddress = connection.ServerUri,
+                DefaultRequestHeaders =
+                {
+                    Authorization = AuthenticationHeaderFactory.Create(
+                        connection.UserName, connection.Password, connection.Authentication),
+                },
+            };
+
+            httpClient.DefaultRequestHeaders.Add("User-Agent", userAgent);
+
+            IsConnected = true;
+
+            var versionResponse = await InvokeRequestAsync<IGetVersionRequest, string>(token);
+            sonarQubeVersion = Version.Parse(versionResponse);
+
+            var credentialResponse = await InvokeRequestAsync<IValidateCredentialsRequest, bool>(token);
+            if (!credentialResponse)
+            {
+                IsConnected = false;
+                throw new InvalidOperationException("Invalid credentials");
+            }
+        }
+
+        public void Disconnect()
+        {
+            IsConnected = false;
+            httpClient.Dispose();
+        }
+
+        private void EnsureIsConnected()
```

Comment:
"CheckIsConnected" might be better - to me "Ensure" suggests that the method will open a connection if not already connected, not that it will throw.

Reply:
On a second thought... `Ensure*` seems in sync with `HttpResponseMessage.EnsureSuccessStatusCode()`, which also throws when the state of the object is not expected.